trio/_core/_result.py

```diff
@@ -1,125 +1,48 @@
-
-           x = fn(*args)
-           x = Result.capture(fn, *args).unwrap()
-
-        """
-
-    @abc.abstractmethod
-    def send(self, gen):
-        """Send or throw the contained value or exception into the given
-        generator object.
-
-        Args:
-          gen: A generator object supporting ``.send()`` and ``.throw()``
-              methods.
+from .. import _deprecate
 
-        """
-
-    @abc.abstractmethod
-    async def asend(self, agen):
-        """Send or throw the contained value or exception into the given async
-        generator object.
-
-        Args:
-          agen: An async generator object supporting ``.asend()`` and
-              ``.athrow()`` methods.
-
-        """
-
-
-@attr.s(frozen=True, repr=False)
-class Value(Result):
-    """Concrete :class:`Result` subclass representing a regular value.
-
-    """
-
-    value = attr.ib()
-    """The contained value."""
-
-    def __repr__(self):
-        return "Value({!r})".format(self.value)
-
-    def unwrap(self):
-        return self.value
-
-    def send(self, gen):
-        return gen.send(self.value)
-
-    async def asend(self, agen):
-        return await agen.asend(self.value)
-
-
-@attr.s(frozen=True, repr=False)
-class Error(Result):
-    """Concrete :class:`Result` subclass representing a raised exception.
-
-    """
-
-    error = attr.ib(validator=attr.validators.instance_of(BaseException))
-    """The contained exception object."""
-
-    def __repr__(self):
-        return "Error({!r})".format(self.error)
-
-    def unwrap(self):
-        raise self.error
-
-    def send(self, it):
-        return it.throw(self.error)
+__all__ = ["Result", "Value", "Error"]
 
-    async def asend(self, agen):
-        return await agen.athrow(self.error)
+_deprecate.enable_attribute_deprecations(__name__)
+
+
+class Result(outcome.Outcome):
+    @classmethod
+    @_deprecate.deprecated(
+        version="0.5.0", issue=494, instead="outcome.capture"
+    )
+    def capture(cls, sync_fn, *args):
+        return outcome.capture(sync_fn, *args)
+
+    @classmethod
+    @_deprecate.deprecated(
+        version="0.5.0", issue=494, instead="outcome.acapture"
+    )
+    async def acapture(cls, async_fn, *args):
+        return await outcome.acapture(async_fn, *args)
+
+
+# alias these types so they don't mysteriously disappear
+Value = outcome.Value
+Error = outcome.Error
+
+# ensure that isinstance(Value(), Result)/issubclass(Value, Result) and etc
+# don't break
+Result.register(Value)
+Result.register(Error)
+
+__deprecated_attributes__ = {
+    "Result":
+        _deprecate.DeprecatedAttribute(
+            Result, "0.5.0", issue=494, instead="outcome.Outcome"
+        ),
+    "Value":
+        _deprecate.DeprecatedAttribute(
+            Value, "0.5.0", issue=494, instead="outcome.Value"
+        ),
+    "Error":
+        _deprecate.DeprecatedAttribute(
+            Error, "0.5.0", issue=494, instead="outcome.Error"
+        )
+}
```

Comment:
Sorry, this `__deprecated_attributes__` thing is pretty confusing, we need some better docs here...

Since it's based on module attribute lookup, using this inside `trio._core._result` means that we'll issue warnings when someone mentions `trio._core._result.<whatever>`, which isn't really what we want – it's references to `trio.hazmat.<whatever>` that people might doing and we want to deprecate. So we want to apply the `__deprecated_attributes__` stuff to `trio.hazmat`. In the version we used to have here we did that in `trio/__init__.py`, which seems as good as anything: https://github.com/python-trio/trio/commit/089c6ff02e624190d4cd3c4f3f2543cabcddfdec#diff-9bcb0dac3a23179589f66a8646fa38faL107

Also, this isn't really relevant because of the previous issue, but as an FYI: I think this also wouldn't work because the `__deprecated_attributes__` stuff is only checked if there isn't a real variable of the same name. (This is an optimization to avoid slowdowns on accessing non-deprecated attributes; it works using `__getattr__`.) So here where are actual variables named `Value`, `Error`, `Result` in the same module, it makes `__deprecated_attributes__` a no-op.